packages/near-api-js/test/providers.test.js

```diff
@@ -295,6 +295,41 @@ test('json rpc light client proof', async() => {
     await expect(provider.lightClientProof(lightClientRequest)).rejects.toThrow(/.+ block .+ is ahead of head block .+/);
 });
 
+test('json rpc get next light client block', withProvider(async (provider) => {
+    const stat = await provider.status();
+
+    // Get block in at least the last epoch (epoch duration 43,200 blocks on mainnet and testnet)
+    const height = stat.sync_info.latest_block_height;
+    const protocolConfig = await provider.experimental_protocolConfig({ finality: 'final' });
+
+    if (height < protocolConfig.epoch_length) {
```

Comment:
I'm a little wary of adding a test like this, branching execution paths based on state extrinsic to the test environment means that for any given test run we are only running an unknown subset of these paths. We should really only execute known code paths to ensure identical, deterministic coverage on every run.

Even as an integration test I'm not convinced there's lots of value to testing the methods that only call `sendJsonRpc` and pass along the parsed response. Typically tests of this nature would be mocked to return known sets of values from RPC to make it a known quantity for any given test. That said this applies to all methods returning `sendJsonRpc` so introducing mocked methods and data doesn't need to be addressed by this PR.

What would you think about repurposing this as a `cookbook` example instead? It's a great demonstration of how to use this method and interpret the logic behind what it's returning.

Reply:
> I'm a little wary of adding a test like this, branching execution paths based on state extrinsic to the test environment means that for any given test run we are only running an unknown subset of these paths. We should really only execute known code paths to ensure identical, deterministic coverage on every run.

Yeah, I wasn't happy with that setup, but I don't think there is a reasonable way to test this code path for _all_ networks. I mean, how tests could be (and maybe should be to be more deterministic) is starting a new network, running the tests, and discarding the data. I could just remove this section, and I believe it passes for all configs this repo uses (testnet, mainnet), but adding different networks for tests like the one mentioned would just cause issues for someone else.

I could just remove this section and return early, essentially skipping the test if we cannot retrieve a block from last epoch, but I figured this would just test to make sure that's the case.
 
> Even as an integration test I'm not convinced there's lots of value to testing the methods that only call `sendJsonRpc` and pass along the parsed response. Typically tests of this nature would be mocked to return known sets of values from RPC to make it a known quantity for any given test. That said this applies to all methods returning `sendJsonRpc` so introducing mocked methods and data doesn't need to be addressed by this PR.

Yeah, unless the provider is mocked, there isn't a way to have a test vector for this, since it's for testnet and mainnet? And if it is mocked, you'd lose the test verifying that the RPC interface matches this client.

> What would you think about repurposing this as a `cookbook` example instead? It's a great demonstration of how to use this method and interpret the logic behind what it's returning.

I don't think this specifically is a good use case since you'd never be doing this specific thing. I'm very open to doing a cookbook example when the validation is done for that flow, but this is not useful on its own.

Happy to just remove this test if that's preferred

Reply:
> I don't think there is a reasonable way to test this code path for all networks

I think you're probably right 😄 

> And if it is mocked, you'd lose the test verifying that the RPC interface matches this client.

Off the top of my head something that mocks `sendJsonRpc` and returns actual RPC responses (i.e. queried from RPC and saved in source as test assets) provides a reasonable canary in the short term, it's just difficult to keep those assets from stagnating as new changes modify them.

> I'm very open to doing a cookbook example when the validation is done for that flow, but this is not useful on its own.

👍 that would be awesome